Userland/Libraries/LibCore/Stream.h

```diff
@@ -0,0 +1,838 @@
+    virtual bool is_open() const override { return m_helper.is_open(); }
+    virtual void close() override { m_helper.close(); }
+    virtual KResultOr<size_t> pending_bytes() const override { return m_helper.pending_bytes(); }
+    virtual KResultOr<bool> can_read_without_blocking(int timeout = 0) const override { return m_helper.can_read_without_blocking(timeout); }
+    KResult set_blocking(bool enabled) override { return m_helper.set_blocking(enabled); }
+    KResult set_close_on_exec(bool enabled) override { return m_helper.set_close_on_exec(enabled); }
+
+    virtual ~UDPSocket() override { close(); }
+
+private:
+    UDPSocket() { }
+
+    void setup_notifier()
+    {
+        VERIFY(is_open());
+
+        m_helper.setup_notifier();
+        m_helper.notifier()->on_ready_to_read = [this] {
+            if (on_ready_to_read)
+                on_ready_to_read();
+        };
+    }
+
+    PosixSocketHelper m_helper { Badge<UDPSocket> {} };
+};
+
+class LocalSocket final : public Socket {
+public:
+    static KResultOr<LocalSocket> connect(String const& path);
+
+    LocalSocket(LocalSocket&& other)
+        : Socket(static_cast<Socket&&>(other))
+        , m_helper(move(other.m_helper))
+    {
+        if (is_open())
+            setup_notifier();
+    }
+
+    LocalSocket& operator=(LocalSocket&& other)
+    {
+        Socket::operator=(static_cast<Socket&&>(other));
+        m_helper = move(other.m_helper);
+        if (is_open())
+            setup_notifier();
+
+        return *this;
+    }
+
+    virtual bool is_readable() const override { return is_open(); }
+    virtual bool is_writable() const override { return is_open(); }
+    virtual KResultOr<size_t> read(Bytes buffer) override { return m_helper.read(buffer); }
+    virtual KResultOr<size_t> write(ReadonlyBytes buffer) override { return m_helper.write(buffer); }
+    virtual bool is_eof() const override { return m_helper.is_eof(); }
+    virtual bool is_open() const override { return m_helper.is_open(); }
+    virtual void close() override { m_helper.close(); }
+    virtual KResultOr<size_t> pending_bytes() const override { return m_helper.pending_bytes(); }
+    virtual KResultOr<bool> can_read_without_blocking(int timeout = 0) const override { return m_helper.can_read_without_blocking(timeout); }
+    KResult set_blocking(bool enabled) override { return m_helper.set_blocking(enabled); }
+    KResult set_close_on_exec(bool enabled) override { return m_helper.set_close_on_exec(enabled); }
+
+    virtual ~LocalSocket() { close(); }
+
+private:
+    LocalSocket() { }
+
+    void setup_notifier()
+    {
+        VERIFY(is_open());
+
+        m_helper.setup_notifier();
+        m_helper.notifier()->on_ready_to_read = [this] {
+            if (on_ready_to_read)
+                on_ready_to_read();
+        };
+    }
+
+    PosixSocketHelper m_helper { Badge<LocalSocket> {} };
+};
+
+// Buffered stream wrappers
+
+template<typename T>
+concept IsStream = IsBaseOf<Stream, T>;
```

Comment:
Concepts are usually named `Foo`, not `IsFoo`.
Though the name is rather unfortunate because it's not exactly a Stream, but something inheriting from a Stream.

I don't have a suggestion for a better name, perhaps a separate namespace could help?

Reply:
If this is not expected to be used outside the class below, perhaps it could be inside that class?

Reply:
Sure, we could put it inside the class, though I can't think of a name that would fit within the class.

Reply:
Just `Stream` should be fine, no?
Can refer to the actual Stream as Core::Stream, etc

If this is internal, the current name sounds...passable, let's just limit it to within the class

Reply:
I guess I could have it like this:

```cpp
class Stream {
public:
    template <typename T>
    concept StreamLike = IsBaseOf<Stream, T>;
};

// ...

template <Stream::StreamLike T>
class Buffered : Core::Stream {
// ...
};
```

What do you think?